opm/autodiff/BlackoilAquiferModel.hpp

```diff
@@ -0,0 +1,131 @@
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <opm/common/utility/platform_dependent/disable_warnings.h>
+#include <opm/common/utility/platform_dependent/reenable_warnings.h>
+
+#include <cassert>
+#include <tuple>
+
+#include <opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp>
+
+#include <opm/parser/eclipse/EclipseState/AquiferCT.hpp>
+#include <opm/parser/eclipse/EclipseState/Aquancon.hpp>
+
+#include <opm/core/simulator/SimulatorReport.hpp>
+
+#include <opm/simulators/timestepping/SimulatorTimer.hpp>
+
+#include <opm/autodiff/BlackoilPropsAdFromDeck.hpp>
+#include <opm/autodiff/BlackoilDetails.hpp>
+#include <opm/autodiff/BlackoilModelParameters.hpp>
+#include <opm/autodiff/RateConverter.hpp>
+#include <opm/autodiff/AquiferCarterTracy.hpp>
+
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/parser/eclipse/Deck/DeckRecord.hpp>
+#include <opm/parser/eclipse/Deck/DeckKeyword.hpp>
+
+#include <dune/common/fmatrix.hh>
+#include <dune/istl/bcrsmatrix.hh>
+#include <dune/istl/matrixmatrix.hh>
+
+#include <opm/material/densead/Math.hpp>
+
+
+namespace Opm {
+
+        /// Class for handling the blackoil well model.
+        template<typename TypeTag>
+        class BlackoilAquiferModel {
+        
+        public:
+
+            // ---------      Types      ---------
+            typedef BlackoilModelParameters ModelParameters;
+
+
+            typedef typename GET_PROP_TYPE(TypeTag, Grid)                Grid;
+            typedef typename GET_PROP_TYPE(TypeTag, GridView)            GridView;
+            typedef typename GET_PROP_TYPE(TypeTag, FluidSystem)         FluidSystem;
+            typedef typename GET_PROP_TYPE(TypeTag, ElementContext)      ElementContext;
+            typedef typename GET_PROP_TYPE(TypeTag, Indices)             BlackoilIndices;
+            typedef typename GET_PROP_TYPE(TypeTag, Simulator)           Simulator;
+            typedef typename GET_PROP_TYPE(TypeTag, Scalar)              Scalar;
+
+            typedef AquiferCarterTracy<TypeTag> Aquifer_object;
+
+            BlackoilAquiferModel(Simulator& ebosSimulator,
+                              const ModelParameters& param,
+                              const bool terminal_output);
+
+            // compute the well fluxes and assemble them in to the reservoir equations as source terms
+            // and in the well equations.
+            void assemble( const SimulatorTimerInterface& timer,
+                           const int iterationIdx                );
+
+            // called at the end of a time step
+            void timeStepSucceeded(const SimulatorTimerInterface& timer);
+
+            inline const Simulator& simulator() const
+            {
+                return ebosSimulator_;
+            }
+
+            // This initialization function is used to connect the parser objects with the ones needed by AquiferCarterTracy
+            void init(const Simulator& ebosSimulator, std::vector<Aquifer_object>& aquifers);
```

Comment:
Both `ebosSimuator` and `aquifers` are member variable, they do not need to function arguments.